osquery/tables/system/windows/prefetch.cpp

```diff
@@ -0,0 +1,427 @@
+  DWORD VolumesInformationSize;
+  FILETIME LastRunTime;
+  DWORD Reserved1[4];
+  DWORD RunCount;
+  DWORD Reserved2;
+} PREFETCH_FILE_INFORMATION, *PPREFETCH_FILE_INFORMATION;
+
+typedef struct _PREFETCH_VOLUME_INFORMATION {
+  DWORD VolumePathOffset;
+  DWORD VolumeDevicePathNumberOfCharacters;
+  FILETIME VolumeCreationTime;
+  DWORD VolumeSerialNumber;
+  DWORD FileReferencesOffset;
+  DWORD FileReferencesDataSize;
+  DWORD DirectoryStringsOffset;
+  DWORD NumberOfDirectoryStrings;
+  DWORD Reserved1;
+} PREFETCH_VOLUME_INFORMATION, *PPREFETCH_VOLUME_INFORMATION;
+
+typedef struct _DIRECTORY_STRING {
+  USHORT Size;
+  WCHAR Directory[1]; // arbitrary size
+} DIRECTORY_STRING, *PDIRECTORY_STRING;
+#pragma pack(pop)
+
+} // namespace
+
+// Convert UTF16 UCHAR vectors to string
+std::string ucharToString(std::string_view data) {
+  size_t data_char = 0;
+  std::string data_string;
+  while (data_char < data.size() && data[data_char] != '\0') {
+    data_string += data[data_char];
+    data_char++;
+    if (data_char == data.size()) {
+      break;
+    }
+    if (data[data_char] == '\0') {
+      data_char++;
+    }
+  }
+  return data_string;
+}
+
+std::vector<std::string> parseAccessedData(std::string_view data_view,
+                                           bool directory) {
+  std::vector<std::string> accessed_data;
+
+  while (data_view.size() > 0) {
+    std::string file_accessed = ucharToString(data_view);
+    if (file_accessed.size() == 0) {
+      break;
+    }
+
+    size_t size = 2;
+    if (directory) {
+      if (file_accessed.size() < 4) {
+        break;
+      }
+
+      size += 2;
+      file_accessed = file_accessed.substr(1);
+    }
+
+    auto check_name = file_accessed.substr(0, 8);
+    // Check if data does not begin with \VOLUME{ or \DEVICE{
+    if (check_name != "\\VOLUME{" && check_name != "\\DEVICE{") {
+      break;
+    }
+
+    accessed_data.push_back(file_accessed);
+    data_view = data_view.substr((file_accessed.size() * 2) + size);
+  }
+  return accessed_data;
+}
+
+PrefetchHeader parseHeader(const PREFETCH_FILE_HEADER* data) {
+  PrefetchHeader result;
+  result.filename = wstringToString(data->FileName);
+  result.prefetch_hash = (boost::format("%x") % data->Hash).str();
+  result.file_size = data->FileSize;
+  return result;
+}
+
+PrefetchFileInfo parseFileInfo(std::string_view data_view) {
+  PrefetchFileInfo result;
+  const auto file_info = (PPREFETCH_FILE_INFORMATION)(
+      &data_view.data()[0] + sizeof(PREFETCH_FILE_HEADER));
+
+  auto size = file_info->FileNameStringsSize;
+  auto offset = file_info->FileNameStringsOffset;
+  if (offset < 0 || offset > data_view.size()) {
+    // Unexpected offset.
+    return result;
+  }
+
+  if (size < 0 || size + offset > data_view.size()) {
+    // Unexpected size.
+    return result;
+  }
+
+  std::vector<std::string> filenames;
+  auto next = (PWCHAR)(&data_view.data()[0] + offset);
+  while (*next != '\0') {
+    auto filename = wstringToString(next);
```

Comment:
@puffyCid and @farfella I don't like this very much because there's a potential to scan past the boundary of the managed buffer `data_view.size()`.

Reply:
Perhaps we should alter the implementation of `wstringToString` to take a max-size.